### PR TITLE
wallet: skip rescan if address batch is empty

### PR DESCRIPTION
**Pull request by benthecarman**

### CODE REVIEW

**Review by guggero — COMMENTED** [review]
This change sounds reasonable to me, though I'm not sure if there's a reason for allowing empty rescans. @Roasbeef, do you have any insights here?

File: wallet/rescan.go [hunk 1]
```diff
@@ -254,13 +254,17 @@ out:
 			log.Infof("Started rescan from block %v (height %d) for %d %s",
 				batch.bs.Hash, batch.bs.Height, numAddrs, noun)
 
-			err := chainClient.Rescan(&batch.bs.Hash, batch.addrs,
-				batch.outpoints)
-			if err != nil {
-				log.Errorf("Rescan for %d %s failed: %v", numAddrs,
-					noun, err)
+			if numAddrs > 0 {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply wallet/rescan.go:257]: It's effectively already skipped, but lower down the call stack:
https://github.com/btcsuite/btcwallet/blob/a792319795563fdcb14e03c6da18ee09fe8b6db9/wallet/rescan.go#L258-L267

[the remark this answers is not in the document]
Comment (benthecarman) [reply wallet/rescan.go:257]: I'm confused, that is the code that was added in this PR

[the remark this answers is not in the document]
Comment (Roasbeef) [reply wallet/rescan.go:257]: LOL, I was viewing the wrong diff 😅

**Comment by losh11** [comment]
Any update?

**Comment by Roasbeef** [comment]
Should have a PR to `lnd` to update the dep to see if anything breaks itest wise. I think only reason to do the rescan is that we get an update that the rescan is finished even if it was empty? So maybe we want to send the finished notification asap if we get an empty batch.

**Comment by litch** [comment]
We're a few days into debugging some performance issues that this would resolve absolutely.

What we're finding is that when running starting up an LND node against mutinynet, the initial chain rescan (for the empty state of the wallet) picks a time ~5k blocks in the past.
```
 Locating suitable block for birthday 2024-04-15 23:13:16 +0000 UTC between blocks 0-1013946
2024-04-17 23:13:20.185 [DBG] LNWL: Checking candidate block: height=506973, hash=000000c080f3d1f9cad4f523469a529a4621e3fb9c3b1e765e462730125c53da, timestamp=2023-10-19 02:04:47 +0000 UTC
...
2024-04-17 23:13:21.224 [DBG] LNWL: Found birthday block: height=1008499, hash=000000fe0d28f5152dc961777e46edf861011d4643069ab9cc3165a750d7fbca, timestamp=2024-04-16 00:22:57 +0000 UTC
```

It consistently picks a block ~5k in the past, and then does a rescan (for an empty wallet).  Under most circumstances that we have found, this rescan takes an extremely long time (~50 minutes).  We have found cases where it is faster than that, but it's pretty consistently slow under a range of conditions.

**Comment by Roasbeef** [comment]
@litch so the height estimation for a wallet's birthday is based on assumption about the block rate interval. If Mutinynet is mining blocks with a diff interval, or one that's variable, then the value will be incorrect. 

With that said, this comment still stands: https://github.com/btcsuite/btcwallet/pull/824#issuecomment-1960486349

If you're running into this issue in production, do you want to take over this PR @litch?

**Comment by guggero** [comment]
Replaced by https://github.com/btcsuite/btcwallet/pull/969 (commit author attribution preserved).

**Pull request closed without merging**